Source/DafnyCore/Verifier/ProofDependency.cs

```diff
@@ -0,0 +1,209 @@
+// Copyright (C) Microsoft Corporation.  All Rights Reserved.
+// Copyright by the contributors to the Dafny Project
+// SPDX-License-Identifier: MIT
+//
+//-----------------------------------------------------------------------------
+using Microsoft.Dafny;
+using IToken = Microsoft.Dafny.IToken;
+using PODesc = Microsoft.Dafny.ProofObligationDescription;
+
+namespace DafnyCore.Verifier;
+
+// A proof dependency represents a particular part of a Dafny program
+// that may be used in the process of proving its correctness. When
+// Boogie proves a particular verification condition, it returns a
+// list of strings, returned by the SMT solver, indicating which
+// program elements were used in completing the proof. After this, the
+// ProofDependencyManager can map each string to a more structured
+// ProofDependency.
+public abstract class ProofDependency {
```

Comment:
An instance of `ProofDependency` seems to be a description of an assumption or assertion in Boogie, together with a token. It serves the purpose of a source location, to tell the user where this assumption or assertion came from, even in cases where it originated from multiple locations, like `CallRequiresDependency`.

I think it would make more sense to use the field `Boogie.IToken tok` in `Absy` to store the 'origin' of a Boogie AST node, including textual description like `requires clause at` if need be.

To a large extend this is already done. `NestedToken` can store multiple locations in a single IToken, and has a `message` field that further explains what this location means.

Here's an example where `NestedToken` used to provide information similar to `CallRequiresDependency`:

```
              // check that the preconditions for the call hold
              foreach (AttributedExpression p in e.Function.Req) {
                Expression precond = Substitute(p.E, e.Receiver, substMap, e.GetTypeArgumentSubstitutions());
                var (errorMessage, successMessage) = CustomErrorMessage(p.Attributes);
                foreach (var ss in TrSplitExpr(precond, etran, true, out var splitHappened)) {
                  if (ss.IsChecked) {
                    var tok = new NestedToken(GetToken(expr), ss.Tok);
```

By using an IToken instead of a ProofDependency to describe an assume/assert, the description is not tied to usage tracking, but can also be used for error reporting, like reporting "for the call at location X, asserting the requires clause at location Y failed".

To summarize, I feel there's duplication between `ProofDependency` and `IToken`. What you could do is use the current `Absy.tok` field to generate the descriptions for the assumes/asserts, and leave tweaking the description of `Absy.tok` for later.

Reply:
Although I can see what you're saying about the duplication, to me it feels like it would be less clear to use an `IToken`. I've always found the nested token notion somewhat awkward, because it doesn't clearly describe the relationship between the outer and inner token. This leads to things like "Related location" in error messages. Those related tokens can have descriptions themselves, but it leads to sort of convoluted messages. I'd instead be inclined to move away from using `NestedToken` (although that would be too substantial an effort to be worthwhile).

Reply:
> I'd instead be inclined to move away from using NestedToken (although that would be too substantial an effort to be worthwhile).

I'm in favor of that, but also for not creating duplication in the meantime. My suggestion is to use NestedToken for now and improve it in a dedicated PR.

Reply:
Moving to NestedToken here would make this code much harder to comprehend, and although the `ProofDependency` type does overlap in purpose with `NestedToken`, it doesn't entirely.

One of the key goals of the current data structure is to allow, as an incremental improvement, embedding more information about the Dafny AST that led to the proof dependency in the first place, to allow for more informative messages. @MikaelMayer has begun doing similar things with `ProofObligationDescription` with very nice results. A `NestedToken` has no semantic content. It just tracks source ranges. Because of the complexity of the Dafny language, and the number of side conditions that come up, it's entirely possible for two source ranges, from two different "dependencies" to entirely overlap. A description string could disambiguate them, but awkwardly. Having the original AST node to disambiguate, instead, seems much more well-structured to me, and makes continually improving messages much more feasible.

Reply:
The next step in this process will be to implement vacuity warnings, and having only `NestedToken` information to determine whether one of the proof dependencies is identical to one of the proof goals would be super fragile. As it is right now, though, it's possible to just look at the `ProofObligationDescripton` inside a `ProofObligationDependency`.